### PR TITLE
task1

### PR DESCRIPTION
**Pull request by andrei-pastukhou**

### CODE REVIEW

File: task1/js/main.js [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    this.ListLinkClasses = config.ListLinkClasses;
+    // Shortcuts to DOM Elements.
+    this.channelListContainer = this.getDomElement('channels_list');
+    this.newsContainer = this.getDomElement('news');
+  }
+
+  // Return dom element with id. If id's element absent create new div element with necessary id.
+  getDomElement(id) {
+    let div = document.getElementById(id);
+    if (! div) {
```
Comment (vladislavkovaliov) [inline task1/js/main.js]:
````
if (!div) { // without space
...
}
````

File: task1/js/main.js [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+      div = document.createElement('div');
+      div.id = id;
```
Comment (vladislavkovaliov) [inline task1/js/main.js]: It difficult for understanding what we are setting. 
Let use 
```
div.setAttribute('id', id);
```
p.s. https://developer.mozilla.org/ru/docs/Web/API/Element/setAttribute

File: task1/js/main.js [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+  'ListLinkClasses' : ['flex-sm-fill', 'text-sm-center', 'nav-link'],  //clases for link in channel choose list
+};
+
+class App {
+
+  constructor() {
+    // Use config like parametr's of app's object.
+    this.API_KEY = config.API_KEY;
+    this.url = config.url;
+    this.channelList = config.channelList;
```
Comment (vladislavkovaliov) [inline task1/js/main.js]: FYI
It would be better jsut inject config object and after that set to instance field.

File: task1/js/main.js [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+  // Do first call of necessary function.
+  start() {
+    this.drawChannelListLink();
+    this.updateNews();
+  }
+
+  // Draw links to choose the new's chanell.
+  drawChannelListLink() {
+    this.channelList.forEach(function (element) {
```
Comment (vladislavkovaliov) [inline task1/js/main.js]: Use arrow function, please. (ES6)

File: task1/js/main.js [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+      let {title, source} = element;
+      let linkText = document.createTextNode(title);
+      let a = document.createElement('a');
+      a.appendChild(linkText);
```
Comment (vladislavkovaliov) [inline task1/js/main.js]: If `a` doesn't set to new value it makes sense use const. 

p.s. Use const as more as possible.

File: task1/js/main.js [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+      }
+      this.channelListContainer.append(a);
+      a.addEventListener('click', function (e) {
+        e.preventDefault();
+        this.clickChannelLink(a);
+      }.bind(this));
+    }.bind(this));
+  }
+
+  // Event of click on channel choose link.
```
Comment (vladislavkovaliov) [inline task1/js/main.js]:
```
/**
 * Call [some event].
 *
 * @param {HTMLElement} a
 */
```

p.s. https://github.com/airbnb/javascript#comments

File: task1/js/main.js [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+  clickChannelLink(a) {
+    let source = a.getAttribute('source');
+    if (localStorage.getItem(source)) {
+      localStorage.removeItem(source);
+      a.classList.remove('active');
```
Comment (vladislavkovaliov) [inline task1/js/main.js]: Use toggle method, please.

p.s. https://developer.mozilla.org/ru/docs/Web/API/Element/classList

File: task1/js/main.js [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+      localStorage.setItem(source, a.innerHTML);
+      a.classList.add('active');
+    }
+    this.updateNews();
+  }
+
+  // Update/redraw list of news.
+  updateNews() {
+    this.newsContainer.innerHTML = '';
+    let sources = [];
```
Comment (vladislavkovaliov) [inline task1/js/main.js]: Use const instead of let, please.

File: task1/js/main.js [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    // Loads all the notes.
+    for (let key in localStorage) {
+      sources.push(key);
+    }
+
+    let source = sources.join(',');
+    if (source) {
+      const url = `${this.url}?sources=${source}&sortBy=publishedAt&apiKey=${this.API_KEY}`;
+      let req = new Request(url);
```
Comment (vladislavkovaliov) [inline task1/js/main.js]: Use const instead of let, please.

File: task1/js/main.js [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+      if(req) {
+        fetch(req)
+          .then(function (response) {
+            return response.json();
+          })
+          .then(function (data) {
+            data.articles.forEach(function (e) {
```
Comment (vladislavkovaliov) [inline task1/js/main.js]: Use arrow function, please. (ES6)
```
data.articles.forEach((article /* It makes sense to set name that can provides info */) => {....}
```

FYI
```
data.articles.forEach(this.drawNews);
```

File: task1/js/main.js [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+  /*
+    Function create new object (NewsNote) and append it to news block.
+    Input is object:
+    {author: 'name of author', description : 'description', publishedAt : 'time when was publish',
+        source : {id: , title},
+     title: 'title of news', url: 'url to source', urlToImage: 'url to get image to news'}
+  */
+  drawNews(news) {
+    if (news) {
+      let note = document.createElement('news-note');
```
Comment (vladislavkovaliov) [inline task1/js/main.js]: Use `const` instead of `let`, please.

File: task1/js/main.js [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+
+  /*
+    Function create new object (NewsNote) and append it to news block.
+    Input is object:
+    {author: 'name of author', description : 'description', publishedAt : 'time when was publish',
+        source : {id: , title},
+     title: 'title of news', url: 'url to source', urlToImage: 'url to get image to news'}
+  */
+  drawNews(news) {
+    if (news) {
```
Comment (vladislavkovaliov) [inline task1/js/main.js]: If `news` is empty array result will be true.

File: task1/js/main.js [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+    this.descriptionElement.textContent = news.description;
+    this.publishedAtElement.textContent = formated_date.format();
+    this.sourceElement.textContent = source_name;
+    this.titleElement.textContent = news.title;
+  }
+}
+
+// Add format function to Data type.
+Date.prototype.format = function (format = 'YYYY-MM-DD hh:mm') {
+  let zeropad = function (number, length) {
```
Comment (vladislavkovaliov) [inline task1/js/main.js]: Use arrow function, please.

It makes sense to define utils functoin for parsing date.

File: task1/js/main.js [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+    formats = {
+      YYYY: this.getFullYear(),
+      MM: zeropad(this.getMonth() + 1),
+      DD: zeropad(this.getDate()),
+      hh: zeropad(this.getHours()),
+      mm: zeropad(this.getMinutes())
+    },
+    pattern = '(' + Object.keys(formats).join(')|(') + ')';
+
+  return format.replace(new RegExp(pattern, 'g'), function (match) {
```
Comment (vladislavkovaliov) [inline task1/js/main.js]: Use arrow function, please.

File: task1/js/main.js [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+Date.prototype.format = function (format = 'YYYY-MM-DD hh:mm') {
+  let zeropad = function (number, length) {
+      number = number.toString();
+      length = length || 2;
+      while (number.length < length)
+        number = '0' + number;
+      return number;
+    },
+    // here you can define your formats
+    formats = {
```
Comment (vladislavkovaliov) [inline task1/js/main.js]: `const formats = {...};`

File: task1/js/main.js [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+      YYYY: this.getFullYear(),
+      MM: zeropad(this.getMonth() + 1),
+      DD: zeropad(this.getDate()),
+      hh: zeropad(this.getHours()),
+      mm: zeropad(this.getMinutes())
+    },
+    pattern = '(' + Object.keys(formats).join(')|(') + ')';
```
Comment (vladislavkovaliov) [inline task1/js/main.js]: `const pattern = {...};`

File: task1/js/main.js [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+    this.sourceElement.textContent = source_name;
+    this.titleElement.textContent = news.title;
+  }
+}
+
+// Add format function to Data type.
+Date.prototype.format = function (format = 'YYYY-MM-DD hh:mm') {
+  let zeropad = function (number, length) {
+      number = number.toString();
+      length = length || 2;
```
Comment (vladislavkovaliov) [inline task1/js/main.js]: What is the magic number 2?

File: task1/js/main.js [hunk 18]
```diff
@@ -0,0 +1,197 @@
+const config = {
```
Comment (vladislavkovaliov) [inline task1/js/main.js]: Please, make choose uppercase, camelCase.

Comment (andrei-pastukhou) [reply task1/js/main.js]: i choosed camelCase. If i understood right, than if name of variable consist of two or more words than name will be like on 'someVariableName'. If name of variable consist of one word than name will be like 'name'.
https://learn.javascript.ru/variable-names (at this page i saw an example)

File: task1/js/main.js (hunk 3 above)
Comment (andrei-pastukhou) [reply task1/js/main.js]: Done.

File: task1/js/main.js (hunk 1 above)
Comment (andrei-pastukhou) [reply task1/js/main.js]: done

File: task1/js/main.js (hunk 2 above)
Comment (andrei-pastukhou) [reply task1/js/main.js]: done

File: task1/js/main.js (hunk 4 above)
Comment (andrei-pastukhou) [reply task1/js/main.js]: done

File: task1/js/main.js (hunk 5 above)
Comment (andrei-pastukhou) [reply task1/js/main.js]: done

File: task1/js/main.js (hunk 6 above)
Comment (andrei-pastukhou) [reply task1/js/main.js]: done

Comment (andrei-pastukhou) [reply task1/js/main.js]: done

File: task1/js/main.js (hunk 7 above)
Comment (andrei-pastukhou) [reply task1/js/main.js]: done

File: task1/js/main.js (hunk 8 above)
Comment (andrei-pastukhou) [reply task1/js/main.js]: done

File: task1/js/main.js (hunk 9 above)
Comment (andrei-pastukhou) [reply task1/js/main.js]: done

File: task1/js/main.js (hunk 10 above)
Comment (andrei-pastukhou) [reply task1/js/main.js]: done

File: task1/js/main.js (hunk 12 above)
Comment (andrei-pastukhou) [reply task1/js/main.js]: done

File: task1/js/main.js (hunk 11 above)
Comment (andrei-pastukhou) [reply task1/js/main.js]: done

File: task1/js/main.js (hunk 13 above)
Comment (andrei-pastukhou) [reply task1/js/main.js]: done 50\50. Already use arrow function but can you explain or give a link, what means utils function?

File: task1/js/main.js (hunk 17 above)
Comment (andrei-pastukhou) [reply task1/js/main.js]: done. define a const and add a comment.
// Minimum amount of character in date and time. For example const = 1 then 8 hour; const = 2 then 08 hour;
      const minNumOfCharacter = 2;

File: task1/js/main.js (hunk 15 above)
Comment (andrei-pastukhou) [reply task1/js/main.js]: done

File: task1/js/main.js (hunk 16 above)
Comment (andrei-pastukhou) [reply task1/js/main.js]: done

File: task1/js/main.js (hunk 15 above)
Comment (andrei-pastukhou) [reply task1/js/main.js]: done

File: task1/js/main.js (hunk 14 above)
Comment (andrei-pastukhou) [reply task1/js/main.js]: done

**Pull request merged**